staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go

```diff
@@ -0,0 +1,366 @@
+
+	"k8s.io/kube-apiextensions-server/pkg/apis/apiextensions"
+	listers "k8s.io/kube-apiextensions-server/pkg/client/listers/apiextensions/internalversion"
+	"k8s.io/kube-apiextensions-server/pkg/registry/customresourcestorage"
+)
+
+// apisHandler serves the `/apis` endpoint.
+// This is registered as a filter so that it never collides with any explictly registered endpoints
+type customResourceHandler struct {
+	versionDiscoveryHandler *customResourceVersionDiscoveryHandler
+	groupDiscoveryHandler   *customResourceGroupDiscoveryHandler
+
+	storageMutationLock sync.Mutex
+	// customStorage contains a map[types.UID]*customResourceInfo
+	customStorage atomic.Value
+
+	requestContextMapper apirequest.RequestContextMapper
+
+	customResourceLister listers.CustomResourceLister
+
+	delegate          http.Handler
+	restOptionsGetter generic.RESTOptionsGetter
+	admission         admission.Interface
+}
+
+// customResourceInfo stores enough information to serve the storage for the custom resource
+type customResourceInfo struct {
+	storage      *customresourcestorage.REST
+	requestScope handlers.RequestScope
+}
+
+func NewCustomResourceHandler(
+	versionDiscoveryHandler *customResourceVersionDiscoveryHandler,
+	groupDiscoveryHandler *customResourceGroupDiscoveryHandler,
+	requestContextMapper apirequest.RequestContextMapper,
+	customResourceLister listers.CustomResourceLister,
+	delegate http.Handler,
+	restOptionsGetter generic.RESTOptionsGetter,
+	admission admission.Interface) *customResourceHandler {
+	ret := &customResourceHandler{
+		versionDiscoveryHandler: versionDiscoveryHandler,
+		groupDiscoveryHandler:   groupDiscoveryHandler,
+		customStorage:           atomic.Value{},
+		requestContextMapper:    requestContextMapper,
+		customResourceLister:    customResourceLister,
+		delegate:                delegate,
+		restOptionsGetter:       restOptionsGetter,
+		admission:               admission,
+	}
+
+	ret.customStorage.Store(map[types.UID]*customResourceInfo{})
+	return ret
+}
+
+func (r *customResourceHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	ctx, ok := r.requestContextMapper.Get(req)
+	if !ok {
+		http.Error(w, "missing context", http.StatusInternalServerError)
+		return
+	}
+	requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+	if !ok {
+		http.Error(w, "missing requestInfo", http.StatusInternalServerError)
+		return
+	}
+	if !requestInfo.IsResourceRequest {
+		pathParts := splitPath(requestInfo.Path)
+		// only match /apis/<group>/<version>
+		if len(pathParts) == 3 {
+			r.versionDiscoveryHandler.ServeHTTP(w, req)
+			return
+		}
+		// only match /apis/<group>
+		if len(pathParts) == 2 {
+			r.groupDiscoveryHandler.ServeHTTP(w, req)
+			return
+		}
+
+		r.delegate.ServeHTTP(w, req)
+		return
+	}
+	if len(requestInfo.Subresource) > 0 {
+		http.NotFound(w, req)
+		return
+	}
+
+	customResourceName := requestInfo.Resource + "." + requestInfo.APIGroup
+	customResource, err := r.customResourceLister.Get(customResourceName)
```

Comment:
disregard my comment above about Get'ability. above we have to see all group versions for a given group. So this does not help there.